crates/sc-proof-of-time/src/pot_state.rs

```diff
@@ -0,0 +1,316 @@
+                    actual: proof.key,
+                });
+            }
+
+            self.proof_of_time
+                .verify(proof)
+                .map_err(PotStateError::InvalidProof)?;
+
+            // All checks passed, advance the tip with the new proof
+            self.add_to_tip(&mut state, proof, ExtendAction::MergeFutureProofs);
+            return Ok(());
+        }
+
+        // Case 3: proof for a future slot
+        self.handle_future_proof(&mut state, &tip, sender, proof)
+    }
+
+    /// Handles the received proof for a future slot.
+    fn handle_future_proof(
+        &self,
+        state: &mut InternalState,
+        tip: &PotProof,
+        sender: PeerId,
+        proof: &PotProof,
+    ) -> Result<(), PotStateError> {
+        // Reject if too much into future
+        if (proof.slot_number - tip.slot_number) > self.config.max_future_slots {
+            return Err(PotStateError::TooFuturistic {
+                tip_slot: tip.slot_number,
+                proof_slot: proof.slot_number,
+            });
+        }
+
+        match state.future_proofs.entry(proof.slot_number) {
+            Entry::Vacant(entry) => {
+                let mut proofs = BTreeMap::new();
+                proofs.insert(sender, proof.clone());
+                entry.insert(proofs);
+                Ok(())
+            }
+            Entry::Occupied(mut entry) => {
+                let proofs_for_slot = entry.get_mut();
+                // Reject if the sender already sent a proof for same slot number.
+                if proofs_for_slot.contains_key(&sender) {
+                    return Err(PotStateError::DuplicateProofFromPeer(sender));
+                }
+
+                // Reject if the there is an existing proof with same
+                // seed/key but different checkpoints.
+                if let Some(existing_proof) =
+                    proofs_for_slot.values().find(|p| p.seed == proof.seed)
+                {
+                    if existing_proof.checkpoints.as_slice() != proof.checkpoints.as_slice() {
+                        return Err(PotStateError::InvalidContextFromPeer(sender));
+                    }
+                }
+
+                // TODO: put a max limit on future proofs per slot number.
+                proofs_for_slot.insert(sender, proof.clone());
+                Ok(())
+            }
+        }
+    }
+
+    /// Called when the chain is extended with a new proof.
+    /// Tries to advance the tip as much as possible, by merging with
+    /// the pending future proofs.
+    fn merge_future_proofs(&self, state: &mut InternalState) {
+        loop {
+            let tip = if let Some(tip) = state.chain.last() {
+                tip.clone()
+            } else {
+                return;
+            };
+
+            // Get the pending proofs for (tip.slot_number + 1).
+            let next_slot = tip.slot_number + 1;
+            let proofs_for_slot = match state.future_proofs.get_mut(&next_slot) {
```

Comment:
> in case PoT chain reorg happens

We don't consider PoT chain reorgs separately from consensus chain reorg - node will only change it's view of PoT chain if it switches to a fork that has different PoT, but never based on gossip . Though there's probably a check to be added to see that your PoT tree extends the PoT from your best block and discard local PoT tree if it doesn't.